### PR TITLE
Analyzer support distributed queries processing

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Planner/Planner.cpp [hunk 1]
```diff
@@ -363,488 +1118,295 @@ void Planner::buildQueryPlanIfNeeded()
...
+                throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                    "Window functions does not support processing from WithMergeableStateAfterAggregation");
+        }
+        else if (expression_analysis_result.hasWindow() || expression_analysis_result.hasAggregation())
+        {
+            if (expression_analysis_result.hasWindow())
+            {
+                const auto & window_analysis_result = expression_analysis_result.getWindow();
+                if (expression_analysis_result.hasAggregation())
+                    addExpressionStep(query_plan, window_analysis_result.before_window_actions, "Before window functions", result_actions_to_execute);
```
Comment (novikd) [inline src/Planner/Planner.cpp:1302]: Haven't we already evaluated it on line 1222?

File: src/Planner/TableExpressionData.h [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 #include <Core/Names.h>
 #include <Core/NamesAndTypes.h>
+#include <Core/ColumnsWithTypeAndName.h>
```
Comment (novikd) [inline src/Planner/TableExpressionData.h:5]: Seems redundant

File: src/Processors/QueryPlan/DistributedCreateLocalPlan.cpp [hunk 3]
```diff
@@ -48,26 +49,48 @@ std::unique_ptr<QueryPlan> createLocalPlan(
...
+            select_query_options);
+
+        interpreter.setProperClientInfo(replica_num, replica_count);
+        if (coordinator)
+        {
+            interpreter.setMergeTreeReadTaskCallbackAndClientInfo([coordinator](PartitionReadRequest request) -> std::optional<PartitionReadResponse>
+            {
+                return coordinator->handleRequest(request);
+            });
+        }
```
Comment (novikd) [inline src/Processors/QueryPlan/DistributedCreateLocalPlan.cpp:89]: Looks like copy-paste. Consider extracting to a separate function.

File: src/Storages/StorageDistributed.cpp [hunk 4]
```diff
@@ -663,12 +691,28 @@ void StorageDistributed::read(
...
+        Planner planner(query_tree_with_replaced_distributed_table, SelectQueryOptions(processed_stage), PlannerConfiguration{.only_analyze = true});
+        planner.buildQueryPlanIfNeeded();
+        header = planner.getQueryPlan().getCurrentDataStream().header;
+    }
+    else
+    {
+        header =
+            InterpreterSelectQuery(query_info.query, local_context, SelectQueryOptions(processed_stage).analyze()).getSampleBlock();
+        query_ast = query_info.query;
+    }
```
Comment (novikd) [inline src/Storages/StorageDistributed.cpp:711]: Consider extracting a separate method

**Review by novikd — COMMENTED** [review]
In general LGTM

File: src/Planner/Planner.h [hunk 5]
```diff
@@ -48,10 +55,15 @@ class Planner
     void addStorageLimits(const StorageLimitsList & limits);
 
 private:
+    void buildUnionNodeQueryPlan();
+
+    void buildQueryNodePlan();
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.h:60]: Updated.

File: src/Planner/Planner.cpp [hunk 6]
```diff
@@ -131,52 +133,792 @@ void checkStoragesSupportTransactions(const PlannerContextPtr & planner_context)
...
+    const PlannerContextPtr & planner_context,
+    const QueryNode & query_node,
+    std::vector<ActionsDAGPtr> & result_actions_to_execute)
+{
+    const auto & query_context = planner_context->getQueryContext();
+    const auto & settings = query_context->getSettingsRef();
+
+    const auto & aggregation_analysis_result = expression_analysis_result.getAggregation();
+    const auto & having_analysis_result = expression_analysis_result.getHaving();
+    bool totals_having_final = !query_node.isGroupByWithRollup() && !query_node.isGroupByWithCube();
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.cpp:386]: Updated.

File: src/Planner/Planner.cpp [hunk 7]
```diff
@@ -131,52 +133,792 @@ void checkStoragesSupportTransactions(const PlannerContextPtr & planner_context)
...
+
+    auto distinct_step = std::make_unique<DistinctStep>(
+        query_plan.getCurrentDataStream(),
+        limits,
+        limit_hint_for_distinct,
+        column_names,
+        pre_distinct,
+        settings.optimize_distinct_in_order);
+
+    if (pre_distinct)
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.cpp:473]: Updated.

File: src/Planner/Planner.cpp [hunk 8]
```diff
@@ -131,52 +133,792 @@ void checkStoragesSupportTransactions(const PlannerContextPtr & planner_context)
...
+                auto expression_to_interpolate_expression_nodes
+                    = planner_actions_visitor.visit(interpolate_actions_dag, interpolate_node_typed.getExpression());
+                auto interpolate_expression_nodes
+                    = planner_actions_visitor.visit(interpolate_actions_dag, interpolate_node_typed.getInterpolateExpression());
+
+                if (expression_to_interpolate_expression_nodes.size() != 1)
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Expression to interpolate expected to have single action node");
+
+                if (interpolate_expression_nodes.size() != 1)
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Interpolate expression expected to have single action node");
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.cpp:590]: Updated.

File: src/Planner/TableExpressionData.h (hunk 2 above)
Comment (kitaisreal) [reply src/Planner/TableExpressionData.h:5]: Updated.

File: src/Planner/Planner.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Planner/Planner.cpp:1302]: We evaluated it on 1222 only if there were no aggregation. If there is aggregation we must finish aggregation on initiator and then evaluate expressions before window functions and windows.

File: src/Planner/Planner.cpp [hunk 9]
```diff
@@ -131,52 +133,792 @@ void checkStoragesSupportTransactions(const PlannerContextPtr & planner_context)
...
+
+        limit_length += limit_offset;
+        limit_offset = 0;
+    }
+
+    const auto & query_context = planner_context->getQueryContext();
+    const Settings & settings = query_context->getSettingsRef();
+
+    auto limit = std::make_unique<LimitStep>(query_plan.getCurrentDataStream(), limit_length, limit_offset, settings.exact_rows_before_limit);
+    if (do_not_skip_offset)
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.cpp:674]: Updated.

File: src/Planner/Planner.cpp [hunk 10]
```diff
@@ -131,52 +133,792 @@ void checkStoragesSupportTransactions(const PlannerContextPtr & planner_context)
...
+    UInt64 limit_length = 0;
+    UInt64 limit_offset = 0;
+};
 
-        Planner subquery_planner(
-            subquery_node,
-            subquery_options,
-            planner_context->getGlobalPlannerContext());
-        subquery_planner.buildQueryPlanIfNeeded();
+void addExpressionStep(QueryPlan & query_plan,
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.cpp:199]: Will change it in separe pull request.

File: src/Planner/Planner.cpp [hunk 11]
```diff
@@ -131,52 +133,792 @@ void checkStoragesSupportTransactions(const PlannerContextPtr & planner_context)
...
+
+                if (interpolate_expression_nodes.size() != 1)
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Interpolate expression expected to have single action node");
+
+                const auto * expression_to_interpolate = expression_to_interpolate_expression_nodes[0];
+                const auto & expression_to_interpolate_name = expression_to_interpolate->result_name;
+
+                const auto * interpolate_expression = interpolate_expression_nodes[0];
+                if (!interpolate_expression->result_type->equals(*expression_to_interpolate->result_type))
+                {
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.cpp:597]: Updated.

File: src/Planner/Planner.cpp [hunk 12]
```diff
@@ -131,52 +133,792 @@ void checkStoragesSupportTransactions(const PlannerContextPtr & planner_context)
...
+    bool is_remote_storage = false;
+
+    const auto & table_expression_node_to_data = planner_context->getTableExpressionNodeToData();
+    if (table_expression_node_to_data.size() == 1)
+    {
+        auto it = table_expression_node_to_data.begin();
+        is_remote_storage = it->second.isRemote();
+    }
+
+    SortDescription group_by_sort_description;
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.cpp:357]: This is related to aggregation in order, and this should be implemented on top of query plan.

File: src/Storages/StorageDistributed.cpp (hunk 4 above)
Comment (kitaisreal) [reply src/Storages/StorageDistributed.cpp:711]: I will try to refactor this function in next pull request, where I will add JOIN support.

File: src/Planner/Planner.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (novikd) [reply src/Planner/Planner.cpp:357]: Let's add a comment about it.

[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/Planner.cpp:358]: Will add in separate pull request.

**Pull request merged**